### ISSUE TITLE
postgres initdb doesn't work from command line in nix-shell

### ISSUE THREAD

**Issue by Aminion** [NixOS/nixpkgs#82732]
**Describe the bug**
Сalling initdb command leads to error

Command call log:
`initdb: file "/nix/store/ig7v414akarh5585a3n311s1g2djfahg-build-tools-wrapper/share/postgresql/postgres.bki" does not exist
This might mean you have a corrupted installation or identified
the wrong directory with the invocation option -L.
`

It seems that  [this](https://github.com/NixOS/nixpkgs/issues/23655) bug is still actual, but for nix-shell, not NixOS itself

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.3.0-40-generic, Ubuntu, 19.10 (Eoan Ermine)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels([user]): `"nixpkgs-20.03pre212770.cc1ae9f21b9"`
 - nixpkgs: `/home/[user]/.nix-defexpr/channels/nixpkgs`

**Comment by bouk** [comment]
I ran into this issue, and figured out it's because of this: https://github.com/postgres/postgres/blob/eab6e460e00d61e5aae95298a0fe8c79a1653184/src/port/path.c#L514-L537

postgres assumes that all the share files and such are relative to the installation location, and it uses the binary name to find that path. We could patch in a `goto no_match;` to make it always use the absolute path, which would do away with the need to link in all the directories.

**Comment by bouk** [comment]
Hmm looking into it some more, I don't think this is fixable without breaking extension support, since those need to be installed into the /share directory.

So the alternative is that you link postgres's /share directory into your env, or provide the -L flag

**Comment by eraserhd** [comment]
Even providing -L does not work, since apparently other scripts try to do the same things:

```
ᐅ initdb -L /nix/store/ic0dc8iy6g880wjp92wrakkaln5k92iq-postgresql-14.6/share/postgresql tmp
The files belonging to this database system will be owned by user "jfelice".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.UTF-8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory tmp ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 20
selecting default shared_buffers ... 400kB
selecting default time zone ... America/New_York
creating configuration files ... ok
running bootstrap script ... 2023-02-08 12:54:42.599 EST [65025] FATAL:  could not open directory "/run/current-system/sw/share/postgresql/timezonesets": No such file or directory
2023-02-08 12:54:42.599 EST [65025] HINT:  This may indicate an incomplete PostgreSQL installation, or that the file "/run/current-system/sw/bin/postgres" has been moved away from its proper location.
child process exited with exit code 1
initdb: removing contents of data directory "tmp"
```

**Comment by eraserhd** [comment]
A work-around:

```
$(readlink -f $(which initdb)) tmp
```

**Comment by chreekat** [comment]
This appears fixed for me on NixOS 23.05. (Ironically, #23655 seems to have regressed for me)

```
[b@kuusi:~/Projects/dotfiles/nixos]$ ns -p postgresql
[nix-shell:~/Projects/dotfiles/nixos]$ which pg_ctl
/nix/store/a0hcylfjxz0qbq2z90bzs1ps6940y2b0-postgresql-14.8/bin/pg_ctl

[nix-shell:~/HaskellFoundation]$ make -f db.makefile
/nix/store/a0hcylfjxz0qbq2z90bzs1ps6940y2b0-postgresql-14.8/bin/pg_ctl initdb -D /home/b/HaskellFoundation/.postgres-work
The files belonging to this database system will be owned by user "b".
This user must also own the server process.

The database cluster will be initialized with locale "en_GB.UTF-8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /home/b/HaskellFoundation/.postgres-work ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... Europe/Helsinki
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

initdb: warning: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    /nix/store/a0hcylfjxz0qbq2z90bzs1ps6940y2b0-postgresql-14.8/bin/pg_ctl -D /home/b/HaskellFoundation/.postgres-work -l logfile start

pg_isready -q || /nix/store/a0hcylfjxz0qbq2z90bzs1ps6940y2b0-postgresql-14.8/bin/pg_ctl start -w -o "-F -h '' -k /home/b/HaskellFoundation/.postgres-work" -l /home/b/HaskellFoundation/.postgres-work/logfile
waiting for server to start.... done
server started
psql -lqt | cut -f1 -d \| | grep -qw ghc_perf || createdb ghc_perf

[nix-shell:~/HaskellFoundation]$

```
